### DISCUSSION TITLE
Time consumption of vips_image_write_to_buffer

### DISCUSSION THREAD

**Discussion by rebornfz** [libvips/libvips#2162] [Q&A] [answered]
Hello！ 
I use vips_image_write_to_buffer() api write a image(about a few hundred kb) to buffer, and it consumes hundreds of ms.
It takes too long. 

The steps of my project are as follows：
1. Download picture(It takes tens of ms)
2. Read from the buffer and render the pictuer(It takes tens of ms)
3. use vips_image_write_to_buffer() write picture(VipsImage) to the buffer(It takes hundreds of ms)
4. send the picture to server(It takes tens of ms)

Hundreds of ms takes too long. How can i shorten this time？

**Comment by jcupitt** [comment]
Hi @rebornfz,

You'll need to make a benchmark and figure out where the time is going.

If you can make a complete, runnable program that shows the type of operations you are performing, I could take a look.

**Reply by rebornfz** [reply]
It is an example.

`

#include <stdio.h>
#include<sys/time.h>
#include <vips/vips.h>
#include <stdlib.h>
#include <math.h>

int main ()
{
  VipsImage *bg;
  VipsImage *good;
  VipsImage *goodarea;
  VipsImage *decorate;
  struct timeval t1, t2, t3, t4;

  gettimeofday(&t1,NULL); 
  //printf("end.tv_sec:%d\n",t1.tv_sec);
  bg = vips_image_new_from_file("background.png", NULL);
  goodarea = vips_image_new_from_file("good_area.png", NULL);
  good = vips_image_new_from_file("good1.png", NULL);
  decorate = vips_image_new_from_file("decorate_1.png", NULL);  
  vips_composite2(bg, goodarea, &bg,VIPS_BLEND_MODE_OVER,"x", 0, "y",0, NULL);
  vips_composite2(bg, good, &bg, VIPS_BLEND_MODE_OVER,"x",0, "y",0, NULL);
  vips_composite2(bg, decorate, &bg, VIPS_BLEND_MODE_OVER, "x", 697, "y", 0, NULL);

  gettimeofday(&t2,NULL);

  void *buff;
  size_t len;
  vips_image_write_to_buffer(bg, ".png", &buff, &len, NULL );

  gettimeofday(&t3,NULL);

  vips_image_write_to_file(bg, "example.png", NULL );

  gettimeofday(&t4,NULL);

  float time_use = time_use=(t2.tv_sec-t1.tv_sec)*1000000+(t2.tv_usec-t1.tv_usec);
  printf("read and render time_use is %.10f\n",time_use);

  time_use = time_use=(t3.tv_sec-t2.tv_sec)*1000000+(t3.tv_usec-t2.tv_usec);
  printf("write to buffer time_use is %.10f\n",time_use);

  time_use = time_use=(t4.tv_sec-t3.tv_sec)*1000000+(t4.tv_usec-t3.tv_usec);
  printf("write to file time_use is %.10f\n",time_use);

  return 0;
}

`

The picture used in example
good_area.png
![decorate_1](https://user-images.githubusercontent.com/27042311/112474626-d73d2f00-8daa-11eb-8162-e03f6a04392c.png)

good1.png
![good1](https://user-images.githubusercontent.com/27042311/112474630-d86e5c00-8daa-11eb-8f71-b4062c3d7bb4.png)

background.png
![background](https://user-images.githubusercontent.com/27042311/112474634-d906f280-8daa-11eb-9529-793114ba6939.png)

decorate_1.png
![good_area](https://user-images.githubusercontent.com/27042311/112474635-d99f8900-8daa-11eb-814c-cd898b6ceedf.png)


My running result：
read and render time_use is 7118.0000000000
write to buffer time_use is 215597.0000000000
write to file time_use is 146854.0000000000

**Comment by jcupitt** [comment]
libvips is a lazy library, so all processing is delayed and happens during `write_to_buffer`.

You can make a few small changes to your program to speed it up:

- use sequential mode
- avoid PNG if possible, it's extremely slow
- you can do your three composite operations in one step using `vips_composite()`

**Reply by rebornfz** [reply]
This is just a very simple example. I need the png and vips_composite() is used in my project. what I need is to create a format picture and send it to server. 
Is there any other way to accomplish this?  How can I use sequential mode?

**Reply by jcupitt** [reply]
But do you need PNG for all these files? If `bg.png` doesn't change often, for example, perhaps you could use vips or TIFF format instead? 

I'll make you can example.

**Reply by rebornfz** [reply]
I get a picture(VipsImage*) after the picture is rendered and I need to send the picture to the server. Then I use vips_image_write_to_buffer to get a format picture that can be sent to the server. Can I use other methods?

**Reply by rebornfz** [reply]
Yes, all pictures are in png format.

**Comment by rebornfz** [comment]
The key steps in my project are as follows;
1.Download png into buffer(http)
2.Read the png from buffer (use vips_image_new_from_buffer) and get the VipsImage *
3.Render the png
4.Get a format picture from VipsImage * (use vips_image_write_to_buffer)
5.Send the picture to server


Is there a problem with my method？

**Comment by jcupitt** [comment]
Here's a version of your benchmark with better timing code:

```C
/* compile with:
 *
 *      gcc -g -Wall comp-bench.c $(pkg-config vips --cflags --libs)
 */

#include <stdio.h>
#include <vips/vips.h>

int
main()
{
        VipsImage *bg;
        VipsImage *good;
        VipsImage *goodarea;
        VipsImage *decorate;

        VIPS_INIT(NULL);

        gint64 start, end;

        start = g_get_monotonic_time();

        bg = vips_image_new_from_file("background.png", NULL);
        goodarea = vips_image_new_from_file("good_area.png", NULL);
        good = vips_image_new_from_file("good1.png", NULL);
        decorate = vips_image_new_from_file("decorate_1.png", NULL);
        vips_composite2(bg, goodarea, &bg,VIPS_BLEND_MODE_OVER,"x", 0, "y",0, NULL);
        vips_composite2(bg, good, &bg, VIPS_BLEND_MODE_OVER,"x",0, "y",0, NULL);
        vips_composite2(bg, decorate, &bg, VIPS_BLEND_MODE_OVER, "x", 697, "y", 0, NULL);
                
        vips_image_write_to_file(bg, "example.png", NULL );
                
        end = g_get_monotonic_time();
                
        printf("took %gs\n", (end - start) / 1e6 );
                
        return 0;
}       
```

I see:

```
john@banana ~/try $ ./a.out 
took 0.090305s
```

**Reply by rebornfz** [reply]
On my pc
It takes 0.21171s. (my poor pc)

**Comment by jcupitt** [comment, marked as answer]
And here's a version using sequential mode and a single call to composite:

```C
/* compile with:
 *
 *      gcc -g -Wall comp-bench2.c $(pkg-config vips --cflags --libs)
 */

#include <stdio.h>
#include <vips/vips.h>

int
main()
{
        VipsImage *bg;
        VipsImage *good;
        VipsImage *goodarea;
        VipsImage *decorate;
        VipsImage *out;

        VipsImage *layers[4];
        VipsArrayInt *x_array;
        VipsArrayInt *y_array;
        int mode[3];
        int x[3];
        int y[3];

        VIPS_INIT(NULL);

        gint64 start, end;

        start = g_get_monotonic_time();

        bg = vips_image_new_from_file("background.png",
                "access", VIPS_ACCESS_SEQUENTIAL, NULL);
        goodarea = vips_image_new_from_file("good_area.png",
                "access", VIPS_ACCESS_SEQUENTIAL, NULL);
        good = vips_image_new_from_file("good1.png",
                "access", VIPS_ACCESS_SEQUENTIAL, NULL);
        decorate = vips_image_new_from_file("decorate_1.png",
                "access", VIPS_ACCESS_SEQUENTIAL, NULL);

        layers[0] = bg;

        layers[1] = goodarea;
        mode[0] = VIPS_BLEND_MODE_OVER;
        x[0] = 0;
        y[0] = 0;

        layers[2] = good;
        mode[1] = VIPS_BLEND_MODE_OVER;
        x[1] = 0;
        y[1] = 0;

        layers[3] = decorate;
        mode[2] = VIPS_BLEND_MODE_OVER;
        x[2] = 697;
        y[2] = 0;

        x_array = vips_array_int_new( x, 3 );
        y_array = vips_array_int_new( y, 3 );

        if(vips_composite(layers, &out, 4, mode, 
                "x", x_array, 
                "y", y_array, 
                NULL))
                vips_error_exit(NULL);

        vips_image_write_to_file(out, "example.png", NULL );

        vips_area_unref( VIPS_AREA( x_array ) );
        vips_area_unref( VIPS_AREA( y_array ) );
        g_object_unref( bg );
        g_object_unref( good );
        g_object_unref( goodarea );
        g_object_unref( decorate );
        g_object_unref( out );

        end = g_get_monotonic_time();

        printf("took %gs\n", (end - start) / 1e6 );

        return 0;
}
```

I see:

```
john@banana ~/try $ ./a.out 
took 0.069968s
```

So about 25% faster.

**Reply by rebornfz** [reply]
took 0.167956s on my PC

**Comment by jcupitt** [comment]
If I change it to use TIFF instead of PNG, I see:

```
john@banana ~/try $ ./a.out 
took 0.014653s
```

About 6 times faster. Almost all of your time is being spent in PNG read and write (it is an extremely slow format).

**Reply by rebornfz** [reply]
Almost all of my time is being spent in PNG write. Just the vips_image_write_to_buffer
 I can't change the png format, I may change the PC (lol)

**Reply by jcupitt** [reply]
PNG read is slow as well. Remember that libvips will not read your PNG images until you call `write_to_buffer`.

**Reply by rebornfz** [reply]
Thank you very much!
